KYC Integration - Get User data for API based on the config

ajeety4 opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->
This PR adds a simple functionality/method to get User data for API based on the kyc configuration.


## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->

[Ticket](https://dimagi.atlassian.net/browse/SC-4239)
[Spec](https://docs.google.com/document/d/1Aco77WDhARfc_I8zWZFoA8SccZGN9WVW8Mv3KpUtGig/edit?pli=1&tab=t.0)

Additionally , a [small update](https://github.com/dimagi/commcare-hq/commit/b497195ad236af968b7987db22fdf6a73029d547#diff-6c6aadbad332b1669bb76e30d57b353b7ecf09c50f8192349ae91d33080abc6aR23) was made to the JSON field so that it defaults to list. 
The schema(example) for the said field would as follows
```
[
  {
    'fieldName': 'first_name',
    'mapsTo': 'firstName',
    'source': 'standard'  // this means to get this field from commcare user info. The other option is 'custom' and data would be fetched from user data store defined in config.
  }
]
```
The view (followup work) that will allow to add or edit this config will include the schema validation.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->
Will be created in upcoming ticket.  Here we just developed an API which is not yet consumed.

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Local testing done. Safe change as this is new feature being developed and is not currently used anywhere.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->
Test cases added

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
None

### Migrations
<!-- Delete this section if the PR does not contain any migrations -->
<!-- https://commcare-hq.readthedocs.io/migrations_in_practice.html -->
- [x] The migrations in this code can be safely applied first independently of the code

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

ajeety4 committed:
make api_field_to_user_data_map json field default to list

ajeety4 committed:
get user data from api based on config

ajeety4 committed:
test cases for get user data for API

kaapstorm commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -0,0 +1,39 @@
+    return user_data_for_api
+
+
+def _get_custom_user_data(user, config):
```
I'm concerned that "custom_user_data" here is ambiguous. It could refer generally to data to be sent to the KYC API, or it could refer specifically to the custom user data returned by  `CommCareUser.get_user_data()`.

The other thing is that the reason that some projects are using a custom case type is because KYC is being used for non-users. They could be beneficiaries, or other people involved in the project whose identities need to be confirmed. Maybe rename to something like "_get_source_data()"?

```suggestion
get _get_source_data(source, config):
    """
    Returns a dictionary of source data.

    ``source`` is a CommCareUser or a CommCareCase.
    """
```

ajeety4 replied:
Thanks Norman for pointing this out. I was not aware earlier that KYC can be used for no-users. 
The suggestion sounds good.  Also made change so that source can be Commcare Case for custom case type.

Addressed in https://github.com/dimagi/commcare-hq/pull/35690/commits/5e32f5bb0d7bc9e61ec2d1317449999fb230c465

ajeety4 committed:
source will be a CommcareCase for other case type

other case type can be used for  verification of beneficiaries or non commcare users

ajeety4 committed:
nit: renames exception

ajeety4 committed:
Merge branch 'master' into ay/kyc-integration-get-data-for-api

zandre-eng commented on corehq/apps/integration/kyc/services.py:
```diff
@@ -0,0 +1,41 @@
+    for mapping in config.api_field_to_user_data_map:
+        try:
+            if mapping['source'] == 'standard':
+                user_data_for_api[mapping['fieldName']] = getattr(source, mapping['mapsTo'])
```
Should this not be the other way around? Where we fetch from `fieldName` and then store it as `mapsTo` for the output?

ajeety4 replied:
We could do that too. 
I am thinking this in terms of just mapping (this `api_field` maps to this `hq_field`) if that makes sense. Also based on the  name used for `api_field_to_user_data_map`, it might be better to keep the order as `api_field` maps to `hq_field`.

zandre-eng replied:
Ah okay gotcha. I slightly misread the code the first time and was under the impression that we were simply going to fetch the HQ properties using the names from `mapsTo`. I see now that we are simply storing a mapping of what property should be mapped to which name.
